File: web/app/js/handler/lancamentoProgramadoParcelaHandler.js
'use strict';

app.factory('LancamentoProgramadoParcelaHandler', function() {
    
    var handleList = function(lancamentoProgramadoParcelaList) { 
        if(!lancamentoProgramadoParcelaList) return null;
        return _.map(lancamentoProgramadoParcelaList, function(lancamentoProgramadoParcela) {
            return handle(lancamentoProgramadoParcela);
        });
    };
    
    var handle = function(lancamentoProgramadoParcela) { 
        var lancamentoProgramadoParcelaHandle = {};
        lancamentoProgramadoParcelaHandle.idLancamentoProgramadoParcela = getId(lancamentoProgramadoParcela);
        lancamentoProgramadoParcelaHandle.planoConta = getPlanoConta(lancamentoProgramadoParcela); 
        lancamentoProgramadoParcelaHandle.centroCusto = getCentroCusto(lancamentoProgramadoParcela);
        lancamentoProgramadoParcelaHandle.lancamento = getLancamento(lancamentoProgramadoParcela);
        lancamentoProgramadoParcelaHandle.numero = getNumero(lancamentoProgramadoParcela);
        lancamentoProgramadoParcelaHandle.dataVencimento = getDataVencimento(lancamentoProgramadoParcela);
        lancamentoProgramadoParcelaHandle.valor = getValor(lancamentoProgramadoParcela);
        return lancamentoProgramadoParcelaHandle;
    };
    
    var getId = function(lancamentoProgramadoParcela) {
        return lancamentoProgramadoParcela.idLancamentoProgramadoParcela || null;
    };
    
    var getPlanoConta = function(lancamentoProgramadoParcela) {
        if(!lancamentoProgramadoParcela.planoConta) return lancamentoProgramadoParcela.planoConta;    
        return { idPlanoConta: lancamentoProgramadoParcela.planoConta.idPlanoConta }; 
    };
    
    var getCentroCusto = function(lancamentoProgramadoParcela) {
        if(!lancamentoProgramadoParcela.centroCusto) return lancamentoProgramadoParcela.centroCusto;    
        return { idCentroCusto: lancamentoProgramadoParcela.centroCusto.idCentroCusto }; 
    };
    
    var getLancamento = function(lancamentoProgramadoParcela) {
        if(!lancamentoProgramadoParcela.lancamento) return lancamentoProgramadoParcela.lancamento;
        return { idLancamentoProgramado: lancamentoProgramadoParcela.lancamento.idLancamento }; 
    };
    
    var getNumero = function(lancamentoProgramadoParcela) {
        return lancamentoProgramadoParcela.numero || null;
    };
    
    var getDataVencimento = function(lancamentoProgramadoParcela) {
        return lancamentoProgramadoParcela.dataVencimento || null;
    };
    
    var getValor = function(lancamentoProgramadoParcela) {
        return lancamentoProgramadoParcela.valor || 0;
    };

    return {
        handle: handle,
        handleList: handleList
    }

});

File: src/java/com/portalpostal/dao/handler/LancamentoProgramadoParcelaHandler.java
package com.portalpostal.dao.handler;

import com.portalpostal.model.CentroCusto;
import com.portalpostal.model.Lancamento;
import com.portalpostal.model.LancamentoProgramado;
import com.portalpostal.model.LancamentoProgramadoParcela;
import com.portalpostal.model.PlanoConta;
import java.sql.ResultSet;
import java.sql.SQLException;
import org.sql2o.ResultSetHandler;

public class LancamentoProgramadoParcelaHandler extends GenericHandler implements ResultSetHandler<LancamentoProgramadoParcela> {
    
    public LancamentoProgramadoParcelaHandler() {
        super("lancamento_programado_parcela");
    }
    
    public LancamentoProgramadoParcelaHandler(String table) {
        super(table);
    }

    public LancamentoProgramadoParcela handle(ResultSet result) throws SQLException {
        LancamentoProgramadoParcela lancamentoProgramadoParcela = new LancamentoProgramadoParcela();
        lancamentoProgramadoParcela.setIdLancamentoProgramadoParcela(getInt(result, "idLancamentoProgramadoParcela"));
        lancamentoProgramadoParcela.setPlanoConta(getPlanoConta(result));
        lancamentoProgramadoParcela.setCentroCusto(getCentroCusto(result));
        lancamentoProgramadoParcela.setLancamentoProgramado(getLancamentoProgramado(result));
        lancamentoProgramadoParcela.setLancamento(getLancamento(result));
        lancamentoProgramadoParcela.setNumero(getInt(result, "numero"));
        lancamentoProgramadoParcela.setDataVencimento(getDate(result, "dataVencimento"));
        lancamentoProgramadoParcela.setValor(getDouble(result, "valor"));
        return lancamentoProgramadoParcela;
    }
    
    private PlanoConta getPlanoConta(ResultSet result) throws SQLException {
        if(!existColumn(result, "plano_conta.idPlanoConta")) return null;
        return new PlanoContaHandler().handle(result); 
    }
    
    private CentroCusto getCentroCusto(ResultSet result) throws SQLException {
        if(!existColumn(result, "centro_custo.idCentroCusto")) return null;
        return new CentroCustoHandler().handle(result); 
    }
        
    private LancamentoProgramado getLancamentoProgramado(ResultSet result) throws SQLException {
        if(!existColumn(result, "lancamento_programado.idLancamentoProgramado")) return null;
        if(!existFKValue(result, "lancamento_programado.idLancamentoProgramado")) return null;
        LancamentoProgramado lancamentoProgramado = new LancamentoProgramado();
        lancamentoProgramado.setIdLancamentoProgramado(getInt(result, "idLancamentoProgramadoParcela", "lancamento_programado"));
        return lancamentoProgramado; 
    }
        
    private Lancamento getLancamento(ResultSet result) throws SQLException {
        if(!existColumn(result, "lancamento.idLancamento")) return null;
        if(!existFKValue(result, "lancamento.idLancamento")) return null;
        Lancamento lancamento = new Lancamento();
        lancamento.setIdLancamento(getInt(result, "idLancamento", "lancamento"));
        return new LancamentoHandler().handle(result); 
    }
    
}
